controllers/flinkcluster_updater.go

```diff
@@ -47,31 +49,98 @@ func (updater *_ClusterStatusUpdater) updateClusterStatusIfChanged() error {
 	}
 
 	// Current status recorded in the cluster's status field.
-	var currentStatus = flinkoperatorv1alpha1.FlinkClusterStatus{}
-	updater.observedState.cluster.Status.DeepCopyInto(&currentStatus)
-	currentStatus.LastUpdateTime = ""
+	var oldStatus = flinkoperatorv1alpha1.FlinkClusterStatus{}
+	updater.observedState.cluster.Status.DeepCopyInto(&oldStatus)
+	oldStatus.LastUpdateTime = ""
 
 	// New status derived from the cluster's components.
 	var newStatus = updater.deriveClusterStatus()
 
 	// Compare
-	var changed = updater.isStatusChanged(currentStatus, newStatus)
+	var changed = updater.isStatusChanged(oldStatus, newStatus)
 
 	// Update
 	if changed {
 		updater.log.Info(
-			"Updating status",
-			"current",
+			"Status changed",
+			"old",
 			updater.observedState.cluster.Status,
 			"new", newStatus)
+		updater.createStatusChangeEvents(oldStatus, newStatus)
 		newStatus.LastUpdateTime = time.Now().Format(time.RFC3339)
 		return updater.updateClusterStatus(newStatus)
 	}
 
-	updater.log.Info("No status change", "state", currentStatus.State)
+	updater.log.Info("No status change", "state", oldStatus.State)
 	return nil
 }
 
+func (updater *_ClusterStatusUpdater) createStatusChangeEvents(
+	oldStatus flinkoperatorv1alpha1.FlinkClusterStatus,
+	newStatus flinkoperatorv1alpha1.FlinkClusterStatus) {
+	if oldStatus.Components.JobManagerDeployment.State !=
+		newStatus.Components.JobManagerDeployment.State {
+		updater.createStatusChangeEvent(
+			"JobManager deployment",
+			oldStatus.Components.JobManagerDeployment.State,
+			newStatus.Components.JobManagerDeployment.State)
+	}
+
+	// JobManager service.
+	if oldStatus.Components.JobManagerService.State !=
+		newStatus.Components.JobManagerService.State {
+		updater.createStatusChangeEvent(
+			"JobManager service",
+			oldStatus.Components.JobManagerService.State,
+			newStatus.Components.JobManagerService.State)
+	}
+
+	// TaskManager.
+	if oldStatus.Components.TaskManagerDeployment.State !=
+		newStatus.Components.TaskManagerDeployment.State {
+		updater.createStatusChangeEvent(
+			"TaskManager deployment",
+			oldStatus.Components.TaskManagerDeployment.State,
+			newStatus.Components.TaskManagerDeployment.State)
+	}
+
+	// Job.
+	if oldStatus.Components.Job == nil && newStatus.Components.Job != nil {
```

Comment:
is there a case where old is not nil but new is nil?

Reply:
No, there shouldn't be such a case.